src/classes/CRLP_AbstractRollup_BATCH.cls

```diff
@@ -0,0 +1,247 @@
+     * out of sequence.
+     */
+    protected Map<Id, Integer> parentIdsNotCommittedWithBatch = new Map<Id, Integer>();
+
+    /** @description Simple counters used to track state */
+    protected Integer batchIteration = 0;
+    protected Integer totalRecordsModified = 0;
+
+    /** @description Are there Soft Credits to Contact-Account rollups to execute? */
+    protected Boolean hasAccountSoftCreditRollups = false;
+
+    /** @description Are there Soft Credits to Contact rollups to execute? */
+    protected Boolean hasContactSoftCreditRollups = false;
+
+    /** @description Specified list of Summary (Account, Contact or GAU) Id's to roll up */
+    protected List<Id> recordIds;
+
+    /** @description The Summary Object Type this batch job will process */
+    protected SObjectType summaryObjectType;
+
+    /*******************************************************************************************************
+     * @description Constructs the class, inits the class instance vars used across all CRLP batch jobs
+     */
+    public CRLP_AbstractRollup_BATCH() {
+        this.statefulRollupsById = new Map<Id, List<CRLP_Rollup>>();
+        this.statefulMapOfOppIdsByAcct = new Map<Id, Set<Id>>();
+        this.statefulListOfSavedIds = new List<Id>();
+        initSoftCreditRollupTypes();
+    }
+
+    /*******************************************************************************************************
+     * @description Constructs the class - rollup is for a specific Summary SObject
+     */
+    public CRLP_AbstractRollup_BATCH(SObjectType objType) {
+        this();
+        this.summaryObjectType = objType;
+    }
+
+    /*******************************************************************************************************
+     * @description Constructs the class - Rollup is for a specific set of Summary Object Id's
+     */
+    public CRLP_AbstractRollup_BATCH(List<Id> recordIds) {
+        this();
+        this.summaryObjectType = recordIds[0].getSobjectType();
+        this.recordIds = recordIds;
+    }
+
+    /**
+     * @description Set the hasAccountSoftCreditRollups and hasContactSoftCreditRollups instance vars based on
+     * the current active customizable rollups in the org
+     */
+    private void initSoftCreditRollupTypes() {
+        this.hasContactSoftCreditRollups = CRLP_Rollup_SVC.hasActiveSoftCreditRollupsForSObject(Contact.SObjectType);
+        this.hasAccountSoftCreditRollups = CRLP_Rollup_SVC.hasActiveSoftCreditRollupsForSObject(Account.SObjectType);
+    }
+
+    /**
+     * @description Resequence the Parent (Summary) records to be in the exact order of the detail records
+     * using the defined parent Id
+     * @param parentIds
+     * @param summaryRecords
+     * @return a list of Parent (Summary) SObject records in the proper order
+     */
+    protected List<SObject> resequenceSummaryRecords(List<Id> parentIds, List<SObject> summaryRecords) {
```

Comment:
Had to think about this one for a while. I knew there was a reason I created this to begin with. It has to do with how the Soft Credits are rolled up. The primary batch query on OCR is ordered by AccountId and then ContactId. If rolling up the Contacts portion of this, then technically the ContactId's are not in exact Id order. When querying the actual Contact records however, they will be retrieved in Contact.Id order. Because blocks of records in the batch may be out of sequence, it becomes important to ensure that the rollups use the order of the ContactId's received in the batch execute method rather than the queried order of Contact records. That allow the logic to properly know which Id is the first and last Id in the execute scope. That (specifically the "last") is needed in the rollups logic to know which record might have more data coming in the next batch execute iteration.

Need to document this better.